src/Features/Core/Portable/Diagnostics/CodeAnalysisDiagnosticAnalyzerService.cs

```diff
@@ -34,18 +35,18 @@ private sealed class CodeAnalysisDiagnosticAnalyzerService : ICodeAnalysisDiagno
         private readonly Workspace _workspace;
 
         /// <summary>
-        /// List of projects that we've finished running "run code analysis" on.  Cached results can now be returned for
-        /// these through <see cref="GetLastComputedDocumentDiagnosticsAsync"/> and <see
-        /// cref="GetLastComputedProjectDiagnosticsAsync"/>.
+        /// Mapping of projects to the diagnostics for the projects that we've finished running "run code analysis" on.
+        /// Cached results can now be returned for these through <see cref="GetLastComputedDocumentDiagnostics"/>
+        /// and <see cref="GetLastComputedProjectDiagnostics"/>.
         /// </summary>
-        private readonly ConcurrentSet<ProjectId> _analyzedProjectIds = [];
+        private readonly ConcurrentDictionary<ProjectId, ImmutableArray<DiagnosticData>> _analyzedProjectToDiagnostics = [];
 
```

Comment:
much simpler model.  instead of calling onto the service to compute diagnostics and then calling back in to fetch whatever it had cached when we asked it to compute, we just grab and cache the values ourselves whne we ask it to compute.  THus not needing any sort of coordination.